frontend/app/src/App.tsx

```diff
@@ -921,37 +919,74 @@ export class App extends PureComponent<Props, State> {
-        basePath
+    if (isNullOrUndefined(baseUriParts)) {
+      logWarning(
+        "unable to get baseUriParts. We might not have an active ConnectionManager."
       )
-      const prevPageName =
+      return
+    }
+    const { basePath } = baseUriParts
+
+    // Figure out the page name in the URL path name
+    const prevPageNameInPath = extractPageNameFromPathName(
+      window.location.pathname,
+      basePath
+    )
+
+    // If mainPageName is undefined, we don't even try to figure out
+    // what page is in the URL already
+    let newPagePath: string
+    let prevPageName: string
+    if (notNullOrUndefined(mainPageName)) {
+      prevPageName =
         prevPageNameInPath === "" ? mainPageName : prevPageNameInPath
       // It is important to compare `newPageName` with the previous one encoded in the URL
       // to handle new session runs triggered by URL changes through the `onHistoryChange()` callback,
       // e.g. the case where the user clicks the back button.
       // See https://github.com/streamlit/streamlit/pull/6271#issuecomment-1465090690 for the discussion.
-      if (prevPageName !== newPageName) {
-        const pagePath = isViewingMainPage ? "" : newPageName
-        const queryString = preserveEmbedQueryParams("").toString()
-        const qs = queryString ? `?${queryString}` : ""
 
-        const basePathPrefix = basePath ? `/${basePath}` : ""
+      // If the new page name is not specified, it's the old page name
+      newPageName = newPageName ?? prevPageName
+      newPagePath = newPageName === mainPageName ? "" : newPageName
+    } else if (notNullOrUndefined(newPageName)) {
+      logError("newPageName specified witout providing mainPageName.")
+      return
+    } else {
+      prevPageName = newPageName = newPagePath = prevPageNameInPath
+    }
+
+    // Extract the query string.
+    // The Previous query string we compare to is ALWAYS the one in the URL bar
+    const prevQueryString = window.location.search
+    // and the New one is always the one from app state (with window URL as a fallback)
+    const requestedQueryString = this.getQueryString()
+
+    // If either the page name or the query params have changed, push a new URL to the page history.
+    if (
+      prevPageName !== newPageName ||
+      !areUserURLSearchParamsEqual(prevQueryString, requestedQueryString)
+    ) {
+      const newQueryParams = preserveEmbedQueryParams(
+        requestedQueryString,
+        prevQueryString
+      )
+      const queryString =
+        newQueryParams.size > 0 ? "?" + newQueryParams.toString() : ""
+      const basePathPrefix = basePath ? `/${basePath}` : ""
+      const pageUrl = `${basePathPrefix}/${newPagePath}${queryString}`
 
-        const pageUrl = `${basePathPrefix}/${pagePath}${qs}`
+      window.history.pushState({}, "", pageUrl)
 
-        window.history.pushState({}, "", pageUrl)
+      // If queryString was specified, send a host message
+      if (queryString) {
```

Comment:
I am not 100% certain that this is a correct place to send a `SET_QUERY_PARAM` message to the host. This is partly because I don't fully understand what `hostCommunicationMgr` is all about -- I think it's something to do with Community Cloud, as it doesn't seem to have any effect when running the application locally.

My reason for being uncertain is this: `maybeUpdatePageUrl`, with the changes introduced in this PR, is _only_ ever called after `App.state.queryString` has been updated -- this is true by design, since the method _gets_ `requestedQueryString` from `App.getQueryString()` which in most cases, loads the value from `App.state`. 

There are [some places](https://github.com/streamlit/streamlit/pull/9049/files#diff-1b95d4e0dda1e5928f99c8303599bd811f140fdd40e028fcfca3077f802d6210R1492) where we also send this message _prior_ to updating `App.state.queryString` but [some places](https://github.com/streamlit/streamlit/pull/9049/files#diff-7f6abc01745f6d3e4805e1279acec8c843ba151febe680c576d1497a3fb5a077R105-R112) where we do not. I do not really understand which is correct. When does the host that is on the other side of this _expect_ to receive this message and what behavior does it induce when it receives one?